### ISSUE TITLE
code splitting for plugins

### ISSUE THREAD

**Issue by illume** [kubernetes-sigs/headlamp#1518]
Some features are very big, and the resulting code can make the app take several seconds to load.

Imagine that a feature has to drag in a whole other massive framework, but it's only shown on one page. Now multiply that by 10 pages.

Code splitting would be useful in these cases to only load the code needed by single pages, only when that page loads.

### Acceptance criteria

- Be able to load part of the feature only when a page is loaded that uses that feature.

**Comment by ankitsharma101** [comment]
Hey @illume 
Before starting work, wanted to share a design proposal for review — I dug into the current plugin-loading code and prior related issues/PRs, and I think the scope here is narrower than the issue title suggests. Happy to adjust based on feedback.

Design Proposal: Code Splitting for Plugins (#1518)

Updated after reviewing prior art in #2551 ("Startup is very slow"), which links directly back to #1518, #1520, #1521, and WIP PR #1961. sniok's Nov 2024 comment on #2551 independently identified the exact problem below, before this PR existed — worth citing directly.

Problem restated

Some features are very big, and the resulting code can make the app take several seconds to load. Imagine that a feature has to drag in a whole other massive framework, but it's only shown on one page.

Key finding: plugins are not loaded as ES modules

fetchAndExecutePlugins (frontend/src/plugin/index.ts) fetches each enabled plugin's main.js as text over HTTP, and runPlugin (frontend/src/plugin/runPlugin.ts) executes it with:

ts
const executePlugin = new PrivateFunction(...args, source);
executePlugin(...values);

This is a sandboxed function-body eval, not import() and not a <script> tag. It exists deliberately so each plugin only gets the permission secrets it's allowed to see and can't pollute global scope.

Consequence: cross-plugin splitting already exists — each plugin's main.js is already its own separate network request, isolated from the core app bundle and from other plugins. Webpack/Vite's normal chunk-graph based code splitting doesn't help between plugins, because there's no module graph connecting them to Headlamp core in the first place.

The real gap: splitting within one plugin's own main.js. All of a plugin's dependencies — including ones only needed by a single route it registers — get bundled into that one file, fetched, and eval'd at app startup (Plugins.tsx calls fetchAndExecutePlugins for every enabled plugin in a single useEffect on mount), regardless of whether the user ever visits that route.

Three independent problems, three phases
Phase 0 — Stop eagerly bundling heavy libs into window.pluginLib (smallest, highest-confidence win)

This is visible directly in the file you pasted — frontend/src/plugin/index.ts has static top-level imports of @monaco-editor/react, monaco-editor, and recharts, which are then attached to window.pluginLib so any plugin can use them without bundling its own copy:

ts
import * as ReactMonacoEditor from '@monaco-editor/react';
import * as MonacoEditor from 'monaco-editor';
import * as Recharts from 'recharts';

Because these are static imports, they get pulled into Headlamp's own initial bundle and load on every startup — whether or not any installed plugin uses Monaco or Recharts, and whether or not the current page needs them. This is exactly what sniok flagged on #2551 in Nov 2024 ("Biggest chunk is the monaco editor, which should be lazy loaded... solution here is to remove direct reexports from pluginLib and only provide lazy loaded editor react component. Same thing with recharts.") — it was never resolved. Notably, frontend/vite.config.ts already defines manualChunks for lodash, @mui/material, xterm, and recharts (each gets its own vendor-*.js chunk file), but Monaco is absent from that list despite being the library sniok called out as the single biggest offender. And even for the libraries that are split into their own chunk file, manualChunks only controls which file the code lands in — it doesn't make the import lazy. Because index.ts imports Recharts with a static top-level import, Vite/Rollup still fetches and parses vendor-recharts.js on every initial page load; it's just no longer inlined into the main chunk. The fix still requires converting these to dynamic imports, not just separate chunk files.

Proposed fix: replace the direct re-exports with lazy-loaded proxies, e.g. window.pluginLib.ReactMonacoEditor becomes a React.lazy(() => import('@monaco-editor/react'))-wrapped component, or a getter that dynamic-imports on first access. This requires auditing how plugins currently consume window.pluginLib.MonacoEditor / Recharts (likely as direct component/module references) to make sure the lazy wrapper is a drop-in replacement, since this is a public plugin API surface and a breaking change here affects every published plugin.

This phase needs no changes to plugin loading (fetchAndExecutePlugins, runPlugin) at all — it's isolated to how core's own bundle constructs window.pluginLib. Smallest, most self-contained, most likely to get a quick review.

Phase 1 — Intra-plugin route-level splitting (small, low-risk, opt-in)

Goal: let a plugin author defer a heavy dependency until its route is actually visited, without any change to how Headlamp core discovers or executes plugins.

Correction from an earlier draft of this proposal: Headlamp's frontend build is Vite/Rollup, not webpack — confirmed directly from frontend/vite.config.ts. (plugins/headlamp-plugin/package.json lists ts-loader, a webpack-era dependency, alongside a full set of vite-plugin-* packages and both @storybook/react-vite and @storybook/react-webpack5 — this looks like tooling mid-migration, worth asking illume about directly rather than assuming.) This matters because the mechanism for making a dynamically-imported chunk resolve to the right URL differs: Vite/Rollup doesn't use __webpack_public_path__; it uses the base config option (already set here via base: process.env.PUBLIC_URL) and resolves dynamic import() chunk URLs relative to that, generally via import.meta.url. Getting a plugin's own dynamically-imported chunk to resolve against that plugin's own served directory (rather than Headlamp's own base) needs the equivalent Vite/Rollup-side mechanism — I have not identified the exact config for this and it needs verifying against whatever actually builds a third-party plugin's main.js (see below).

Also unresolved: I could not find the actual bundler config used to build a third-party plugin's main.js output. plugins/headlamp-plugin/package.json's own build script only compiles the headlamp-plugin CLI/library itself via tsc — it doesn't show the bundler invoked when a plugin author runs headlamp-plugin build on their own plugin. That logic likely lives in plugins/headlamp-plugin/bin/ or plugins/headlamp-plugin/config/ (both visible in the repo but not yet inspected). This needs to be checked before Phase 1 can be scoped concretely — everything below is provisional until that's confirmed.

Changes needed (provisional, pending the above):

Whatever bundler headlamp-plugin build actually invokes for third-party plugins: enable chunk splitting for dynamic import(), and ensure the resulting chunk URLs resolve against that plugin's own served path (equivalent to setting output.publicPath in webpack, or Vite/Rollup's base/experimentalRenderBuiltUrl, depending on what's actually in use).
Route.component type (frontend/src/lib/router/Route.ts) — verified: already accepts ExoticComponent<{}> | (() => ReactNode), and React.lazy() returns an ExoticComponent. No type change needed; plugin authors can write component: React.lazy(() => import('./TraceList')) today, as far as the type system is concerned.
Document the pattern for plugin authors in the plugin functionality docs.

No changes required to index.ts, runPlugin.ts, or the fetch/execute/permission-secret model. This is why Phase 1 is worth doing first — it's additive, opt-in per plugin, and doesn't touch the trickiest part of the codebase. But item 1 above needs concrete investigation before this phase can be scoped as more than a sketch.

Phase 2 — Route-level splitting in Headlamp core itself

Separately from plugins: WIP PR #1961 (illume, referenced from #2551) attempted lazy-loading Headlamp's own routes broadly, but was closed unmerged in Aug 2025 after going stale for over a year (auto-closed by the k8s-triage-robot; last real commits were roughly two years ago, and it needed a rebase even before going stale). It's dead, not "in progress" — treat it as prior art to cite and explicitly supersede, not a branch to pick back up.

More importantly, sniok's very first review comment on #1961 pushes back on the premise of the PR itself:

Splitting on every route might be an overkill imho, usually it's only needed in components that import big libraries (for example EditorDialog that imports monaco would be good case for lazy loading). Otherwise it'll just generate a lot of small chunks which might negatively impact build time

This is a core reviewer, over a year before this proposal, independently recommending the targeted approach (split specifically around Monaco-sized dependencies) over blanket per-route splitting. It's direct validation for leading with Phase 0 rather than reviving #1961's broader approach. If Phase 2 is pursued at all, it should be scoped narrowly to components known to pull in large libraries (Monaco, Recharts, anything else identified via bundle analysis), not "every route," to avoid the build-time and chunk-count regression sniok flagged.

Given this, Phase 2 is lower priority than originally scoped — it may turn out to be largely redundant with a well-executed Phase 0 plus per-component lazy loading of specific known-large dialogs/views, rather than a distinct router-level effort.

Phase 3 — Deferred plugin activation (bigger, higher-risk)

Goal: don't fetch/execute a plugin's main.js at all until the user navigates somewhere that plugin cares about.

This is much harder because registerSidebarEntry / registerRoute calls happen inside initialize(), which only runs after a plugin's full source has been fetched and executed. So today there's a chicken-and-egg problem: the sidebar entry a user clicks to reach a page can only exist after the plugin's code has already run.

Possible approaches, roughly in order of increasing scope:

Declarative manifest: extend package.json (already fetched separately, per fetchAndExecutePlugins) with an optional declared list of sidebar entries / routes the plugin will register, so Headlamp can render sidebar entries before the plugin's JS runs, and only fetch+eval main.js on first navigation to one of its routes. Requires a new plugin API surface and a migration path for existing plugins (backwards compat: plugins without a manifest keep loading eagerly, as today).
Split manifest + full bundle: keep runtime registration as-is, but require plugin builds to emit a tiny "route stub" chunk (metadata + lazy-loaded route component) separately from the rest of plugin logic — effectively pushing Phase 1's per-route splitting to be automatic rather than opt-in.

This phase touches the plugin loading/execution/priority code you're already familiar with from index.ts (fetchAndExecutePlugins, applyPluginPriority, filterSources) and has real backwards-compatibility implications for the plugin ecosystem, so it needs its own design discussion and probably its own issue.

Recommendation

Propose Phase 0 as the actual scope of a first PR against #1518 (or against #1520 "code size for plugins", which may be the more accurate issue for this specific fix — worth confirming which one illume wants it filed under). It:

fixes a problem already independently diagnosed and left unresolved since Nov 2024,
is fully self-contained to frontend/src/plugin/index.ts's construction of window.pluginLib,
requires no changes to fetchAndExecutePlugins/runPlugin/the permission-secret model,
benefits every user immediately, not just installations of plugins that happen to use Monaco/Recharts,
can be validated with a before/after bundle-size diff on the main chunk.

Phase 1 (opt-in per-plugin lazy routes) is good follow-up work but depends on confirming headlamp-plugin's build tooling and Route.component typing support it — smaller in code, bigger in coordination since it's a plugin-author-facing API.

Phases 2 and 3 are bigger and shouldn't block Phase 0. Worth filing as explicit follow-ups referencing both #1518 and #2551 so the connection is visible, rather than being silently rediscovered a third time.

Open questions for illume
Is Phase 0 still unaddressed? Resolved: confirmed directly — the current frontend/src/plugin/index.ts still has the static ReactMonacoEditor/MonacoEditor/Recharts imports as of this writing. Not fixed since sniok's Nov 2024 comment.
What's the current status of PR #1961? Resolved: closed unmerged in Aug 2025, stale for over a year. Not salvageable as a branch; useful only as prior art and as the source of sniok's "target big-library components, not every route" recommendation.
Revised: what bundler does headlamp-plugin build actually invoke for third-party plugins (webpack, Vite/Rollup, something else)? The headlamp-plugin package's own package.json didn't show it directly — worth pointing me at plugins/headlamp-plugin/bin/ or plugins/headlamp-plugin/config/ if those hold the answer, since I don't want to guess again given the mixed webpack/Vite dependency signals in that package.json.
Is Route.component already lazy-compatible? Resolved: yes, confirmed from the current type definition — accepts ExoticComponent<{}>, which is what React.lazy() produces.
Should this work be scoped/filed under #1518, #1520 ("code size for plugins"), or split across both depending on phase?
Given sniok's pushback on #1961 (blanket per-route splitting = overkill, target known-large components instead), does it make sense to fold "Phase 2" into Phase 0/1 as a single "target the specific known offenders" effort rather than treating it as separate, router-wide work?
